tests/engine/processes/calcjobs/test_calc_job.py

```diff
@@ -120,43 +121,122 @@ def prepare_for_submission(self, folder):
 class MultiCodesCalcJob(CalcJob):
     """`MultiCodesCalcJob` implementation to test the calcinfo with multiple codes set.
 
-    The codes are run in parallel. The codeinfo1 is set to run without mpi, and codeinfo2
-    is set to run with mpi.
+    The codes are run in parallel and each code enforces a different value for ``withmpi`` on the ``CodeInfo``.
     """
 
     @classmethod
     def define(cls, spec):
         super().define(spec)
-        spec.input('parallel_run', valid_type=orm.Bool)
+        spec.input('code_with_mpi_none', valid_type=orm.AbstractCode, required=False)
+        spec.input('code_with_mpi_true', valid_type=orm.AbstractCode, required=False)
+        spec.input('code_with_mpi_false', valid_type=orm.AbstractCode, required=False)
+        spec.input('parallel_run', valid_type=orm.Bool, default=lambda: orm.Bool(True))
 
     def prepare_for_submission(self, folder):
         from aiida.common.datastructures import CalcInfo, CodeInfo, CodeRunMode
 
-        codeinfo1 = CodeInfo()
-        codeinfo1.code_uuid = self.inputs.code.uuid
-        codeinfo1.withmpi = False
+        calc_info = CalcInfo()
+        calc_info.codes_info = []
 
-        codeinfo2 = CodeInfo()
-        codeinfo2.code_uuid = self.inputs.code.uuid
+        for key, with_mpi in [('code_with_mpi_none', None), ('code_with_mpi_true', True),
+                              ('code_with_mpi_false', False)]:
+            if key in self.inputs:
+                code_info = CodeInfo()
+                code_info.code_uuid = self.inputs[key].uuid
+                if with_mpi is not None:
+                    code_info.withmpi = with_mpi
+                calc_info.codes_info.append(code_info)
 
-        calcinfo = CalcInfo()
-        calcinfo.codes_info = [codeinfo1, codeinfo2]
         if self.inputs.parallel_run:
-            calcinfo.codes_run_mode = CodeRunMode.PARALLEL
+            calc_info.codes_run_mode = CodeRunMode.PARALLEL
         else:
-            calcinfo.codes_run_mode = CodeRunMode.SERIAL
-        return calcinfo
+            calc_info.codes_run_mode = CodeRunMode.SERIAL
+
+        return calc_info
+
+
+@pytest.mark.parametrize(
+    'code_key, with_mpi_code, with_mpi_option, expected', (
+        ('code_with_mpi_none', None, True, 3),
```

Comment:
maybe the name of this input could be `codeinfo_with_mpi_none`, that would make it easier to understand

Reply:
done